### PR TITLE
HPCC-26457 Prevent wuidread resolving result too early.

### PR DESCRIPTION
**Pull request by jakesmith**
Workunit read should not try to get the result when it is
constructed, because conditional branches which aren't used
are still constructed and the result may be from an unexected
dependent subgraph.
Delay checking for result until preparing graph to execute.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review.

This is not an ideal fix, it's essentially moving the check from factory time (which caused the problem) to when the graph is actually used and serialized to slaves.
It really shouldn't be creating the activities that are down conditional branches that aren't [currently] being executed at all, until they're needed, and I tried changing so that was so, but there are too many knock on effects of the way the graphs are initialized/reinitialized and wired together. This is a much more pragmatic fix.

**Comment by jakesmith** [comment]
NB: this issue is ancient - I guess it's quite rare for a wuidread to be in a child query, and dependent on a result, down a conditional arm that isn't executed via some other means.
The issue was hit in 7.6.40-1. I'm currently targetting 8.2.x.

File: thorlcr/master/thactivitymaster.cpp [hunk 1]
```diff
@@ -80,28 +80,82 @@ MODULE_INIT(INIT_PRIORITY_STANDARD)
...
+        e->Release();
+        throw MakeThorException(TE_FailedToRetrieveWorkunitValue, "Failed to find value %s:%d in workunit %s [%s]", stepname?stepname:"(null)", sequence, wuid?wuid:"(null)", text.str());
+    }
+    return false;
+}
+
+
 class CGenericMasterGraphElement : public CMasterGraphElement
 {
+    StringBuffer wuidDiskFilename;
```
Comment (ghalliday) [inline thorlcr/master/thactivitymaster.cpp:116]: Doesn't seem like it is quite in the right place.

Also seems a shame to add it to every activity in the master.  Probably doesn't actually matter.

**Review by ghalliday — COMMENTED** [review]
The change looks ok, but wondering if there is an easy change so that the logic was a bit cleaner.  I suspect the answer is no.

File: thorlcr/master/thactivitymaster.cpp (hunk 1 above)
Comment (jakesmith) [reply thorlcr/master/thactivitymaster.cpp:116]: I could move into a specialized derived class for TAKworkunitread, but wasn't sure it was worth it, or would make it clearer.

**Comment by jakesmith** [comment]
> The change looks ok, but wondering if there is an easy change so that the logic was a bit cleaner. I suspect the answer is no.

I looked at various ways of tackling this.
1) prevent creation of activities down [so far] unused arms of conditionsl, i.e. delay until needed.
But that leads to issues/complications with rewiring the graphs as conditionals change, and also complications with partly started activities like spitter. Solvable, and potentially cleaner in the long run, but would require a lot more changes and more risky, also may not be as effiecient for child queries as they would need to re-wire below conditionals if there were changes.
2) an implementation of an act. that conditionally behaved as a diskread (by using a contained proxy act.), it worked but less clean/more changes.
3) this impl.

This seemed simplest most pragmatic solution to the issue.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**